plugins/amqp_trx_plugin/amqp_trx_plugin.cpp

```diff
@@ -0,0 +1,249 @@
+      "Will consume from 'trx' queue and publish to 'trace' queue.");
+   op("amqp-trx-exchange", bpo::value<std::string>()->default_value(""),
+      "Existing AMQP exchange to send transaction trace messages.");
+   op("amqp-trx-publish-all-traces", bpo::bool_switch()->default_value(false),
+      "If specified then all traces will be published; otherwise only traces for consumed 'trx' queue transactions.");
+}
+
+void amqp_trx_plugin::plugin_initialize(const variables_map& options) {
+   try {
+      my->chain_plug = app().find_plugin<chain_plugin>();
+      EOS_ASSERT( my->chain_plug, chain::missing_chain_plugin_exception, "chain_plugin required" );
+
+      EOS_ASSERT( options.count("amqp-trx-address"), chain::plugin_config_exception, "amqp-trx-address required" );
+      my->amqp_trx_address = options.at("amqp-trx-address").as<std::string>();
+      my->amqp_trx_exchange = options.at("amqp-trx-exchange").as<std::string>();
+      my->amqp_trx_publish_all_traces = options.at("amqp-trx-publish-all-traces").as<bool>();
+   }
+   FC_LOG_AND_RETHROW()
+}
+
+void amqp_trx_plugin::plugin_startup() {
+   handle_sighup();
+   try {
+
+      my->thread_pool.emplace( "amqp_t", 1 );
+
+      my->amqp_trx.emplace( logger, my->thread_pool->get_executor(), my->amqp_trx_address, "trx" );
+      my->amqp_trace.emplace( logger, my->thread_pool->get_executor(), my->amqp_trx_address, "trace" );
+
+      auto& consumer = my->amqp_trx->consume();
+      consumer.onSuccess( []( const std::string& consumer_tag ) {
+         fc_dlog( logger, "consume started: ${tag}", ("tag", consumer_tag) );
+      } );
+      consumer.onError( []( const char* message ) {
+         fc_wlog( logger, "consume failed: ${e}", ("e", message) );
+      } );
+      consumer.onReceived( [my=my](const AMQP::Message& message, uint64_t delivery_tag, bool redelivered) {
+         if( my->shutting_down ) {
+            my->amqp_trx->reject( delivery_tag );
+            return;
+         }
+         if( my->consume_message( message.body(), message.bodySize() ) ) {
+            my->amqp_trx->ack( delivery_tag );
+         } else {
+            my->amqp_trx->reject( delivery_tag );
+         }
+      } );
+
+      if( my->amqp_trx_publish_all_traces ) {
+         auto& chain_plug = app().get_plugin<chain_plugin>();
+         my->applied_transaction_connection.emplace(
+               chain_plug.chain().applied_transaction.connect(
+                     [&]( std::tuple<const chain::transaction_trace_ptr&, const chain::packed_transaction_ptr&> t ) {
+                        my->on_applied_transaction( std::get<0>( t ), std::get<1>( t ) );
+                     } ) );
+      }
+
+   } catch( ... ) {
+      // always want plugin_shutdown even on exception
+      plugin_shutdown();
+      throw;
+   }
+}
+
+void amqp_trx_plugin::plugin_shutdown() {
+   try {
+      fc_dlog( logger, "shutdown.." );
+      my->shutting_down = true; // stop receiving transactions to consume
```

Comment:
The worker thread does not synchronize with anything later in this function, so it's entirely possible that we could return from `plugin_shutdown` and continue on to shutdown the `chain_plugin` all before `shutting_down` becomes visible to the worker thread.

Reply:
`shutting_down` is atomic bool

Reply:
Initial state: `chain_plugin` and `amqp_trx_plugin` are running.
Threads: `app`, `amqp`

I believe that the following order is consistent:

1. app: enter app().shutdown()
2. app: enter amqp_trx_plugin.plugin_shutdown()
3. amqp: read my->shutting_down == true
4. app:  shutting_down = false
5. app: shutdown and destroy other plugins
6. amqp: handle_message
7. app: `my->chain_plug->accept_transaction` (posted from 6)
8. app: `chain_plug` is a dangling pointer.  KABOOM

Note that it appears that this is only possible if `http_plugin` fails to initialize, as everywhere else uses `app().quit()` instead of `app.shutdown()`.

`app.quit()` has a different problem: `plugin_shutdown` is run after the event loop has already exited, so the handler passed to `post` is never run.  This avoids blowing up in `accept_transaction`, but means that the `amqp` thread is joined in the destructor of `amqp_trx_plugin_impl`.  All the members of `amqp_trx_plugin_impl` declared after `thread_pool` will have been destroyed at this point.  KABOOM.